src/joystick/gdk/SDL_gameinputjoystick.cpp

```diff
@@ -0,0 +1,582 @@
+    return 0;
+}
+
+static void GAMEINPUT_JoystickUpdate(SDL_Joystick *joystick)
+{
+    static WORD s_XInputButtons[] = {
+        GameInputGamepadA, GameInputGamepadB, GameInputGamepadX, GameInputGamepadY,
+        GameInputGamepadLeftShoulder, GameInputGamepadRightShoulder, GameInputGamepadView, GameInputGamepadMenu,
+        GameInputGamepadLeftThumbstick, GameInputGamepadRightThumbstick,
+        0 /* Guide button is not supported on Xbox so ignore that... */
+    };
+    Uint8 btnidx = 0, btnstate = 0, hat = 0;
+    GAMEINPUT_InternalJoystickHwdata *hwdata = joystick->hwdata;
+    IGameInputDevice *device = hwdata->devref->device;
+    IGameInputReading *reading = NULL;
+    uint64_t ts = 0;
+    GameInputGamepadState state;
+    HRESULT hR = g_pGameInput->GetCurrentReading(
+        GameInputKindGamepad,
+        device,
+        &reading
+    );
+
+    if (FAILED(hR)) {
+        /* don't SetError here since there can be a legitimate case when there's no reading avail */
+        return;
+    }
+
+    /* GDKX private docs for GetTimestamp: "The microsecond timestamp describing when the input was made." */
+    /* SDL expects a nanosecond timestamp, so I guess US_TO_NS should be used here? */
+    ts = SDL_US_TO_NS(reading->GetTimestamp());
+
+    if (((!hwdata->lastTimestamp) || (ts != hwdata->lastTimestamp)) && reading->GetGamepadState(&state)) {
+        /* `state` is now valid */
+
+#define tosint16(_TheValue) ((Sint16)(((_TheValue) < 0.0f) ? ((_TheValue) * 32768.0f) : ((_TheValue) * 32767.0f)))
+        SDL_SendJoystickAxis(ts, joystick, 0, tosint16(state.leftThumbstickX));
+        SDL_SendJoystickAxis(ts, joystick, 1, tosint16(state.leftThumbstickY));
+        SDL_SendJoystickAxis(ts, joystick, 2, tosint16(state.leftTrigger));
+        SDL_SendJoystickAxis(ts, joystick, 3, tosint16(state.rightThumbstickX));
+        SDL_SendJoystickAxis(ts, joystick, 4, tosint16(state.rightThumbstickY));
+        SDL_SendJoystickAxis(ts, joystick, 5, tosint16(state.rightTrigger));
+#undef tosint16
+
+        for (btnidx = 0; btnidx < (Uint8)SDL_arraysize(s_XInputButtons); ++btnidx) {
+            if (s_XInputButtons[btnidx] == 0) {
+                btnstate = SDL_RELEASED;
+            } else {
+                btnstate = (state.buttons & s_XInputButtons[btnidx]) ? SDL_PRESSED : SDL_RELEASED;
+            }
+
+            SDL_SendJoystickButton(ts, joystick, btnidx, btnstate);
+        }
+
+        if (state.buttons & GameInputGamepadDPadUp) {
+            hat |= SDL_HAT_UP;
+        }
+        if (state.buttons & GameInputGamepadDPadDown) {
+            hat |= SDL_HAT_DOWN;
+        }
+        if (state.buttons & GameInputGamepadDPadLeft) {
+            hat |= SDL_HAT_LEFT;
+        }
+        if (state.buttons & GameInputGamepadDPadRight) {
+            hat |= SDL_HAT_RIGHT;
+        }
+        SDL_SendJoystickHat(ts, joystick, 0, hat);
+
+        /* Xbox doesn't let you obtain the power level, pretend we're always full */
```

Comment:
You can get this from IGameInputDevice::GetBatteryState, but the code needs to be prepared to deal with a battery status of "unknown" for devices where this isn't supported.  (Likely by just returning "full power" as it's doing here.)